cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java

```diff
@@ -152,6 +152,41 @@ default CloseableIterator<StructuredRow> scan(Range keyRange, int limit, SortOrd
    */
   CloseableIterator<StructuredRow> scan(Field<?> index) throws InvalidFieldException, IOException;
 
+  /**
+   * Read a set of rows from the table matching the index and the key range.
+   * The rows returned will be sorted on the primary key order.
+   *
+   * @param keyRange key range for the scan
+   * @param limit maximum number of rows to return
+   * @param filterIndex the index to filter upon
+   * @return a {@link CloseableIterator} of rows
+   * @throws InvalidFieldException if the field is not part of the table schema, or is not an indexed column,
+   *                               or the type does not match the schema
+   * @throws IOException if there is an error scanning the table
+   */
+  default CloseableIterator<StructuredRow> scan(Range keyRange, int limit, Field<?> filterIndex)
+    throws InvalidFieldException,
+    IOException {
+    throw new UnsupportedOperationException("No supported implementation.");
+  }
+
+  /**
+   * Read a set of rows from the table matching the key range, return by sortOrder of specified index field.
+   *
+   * @param keyRange key range for the scan
+   * @param limit maximum number of rows to return
+   * @param orderByField the field to sort upon
+   * @param sortOrder the sort order of the index field
+   * @return a {@link CloseableIterator} of rows
+   * @throws InvalidFieldException if the field is not part of the table schema, or is not an indexed column,
+   *                               or the type does not match the schema
+   * @throws IOException if there is an error scanning the table
+   */
+  default CloseableIterator<StructuredRow> scan(Range keyRange, int limit, String orderByField, SortOrder sortOrder)
+    throws InvalidFieldException, IOException {
+    return scan(keyRange, limit);
```

Comment:
this default implementation would sort by primary key, which is the wrong behavior. It would be better to throw an UnsupportedOperationException like the other new method.

Reply:
yup changed it